pywiscat/wis1/util.py

```diff
@@ -0,0 +1,168 @@
+#
+# Copyright (c) 2021 Government of Canada
+# Copyright (c) 2021 IBL Software Engineering spol. s r. o.
+#
+# Permission is hereby granted, free of charge, to any person
+# obtaining a copy of this software and associated documentation
+# files (the "Software"), to deal in the Software without
+# restriction, including without limitation the rights to use,
+# copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following
+# conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# =================================================================
+
+import logging
+import math
+import os
+from lxml import etree
+
+LOGGER = logging.getLogger(__name__)
+
+NAMESPACES = {
+    'gco': 'http://www.isotc211.org/2005/gco',
+    'gmd': 'http://www.isotc211.org/2005/gmd',
+    'gml': 'http://www.opengis.net/gml/3.2',
+    'gmx': 'http://www.isotc211.org/2005/gmx',
+    'gts': 'http://www.isotc211.org/2005/gts',
+    'xlink': 'http://www.w3.org/1999/xlink'
+}
+
+
+def search_files_by_term(directory: str, terms: list) -> list:
+    """
+    Searches directory tree of metadata for files containing search terms
+
+    :param directory: directory to metadata files
+    :param terms: list of terms
+
+    :returns: list of file names
+    """
+
+    matches = []
+
+    LOGGER.debug(f'Walking directory {directory}')
+    match_count = 0
+    file_count = 0
+    for root, _, files in os.walk(directory):
+        for name in files:
+            filename = f'{root}/{name}'
+            LOGGER.debug(filename)
+
+            file_count += 1
+            processed = math.floor(file_count * 100 / len(files))
+
+            if not filename.endswith('.xml'):
+                continue
+
+            e = etree.parse(filename)
+            anytext = ' '.join(
+                [value.strip() for value in e.xpath('//text()')])
+
+            if all(term.lower() in anytext.lower() for term in terms):
+                match_count += 1
+                LOGGER.debug(f'Found match #{match_count}, searched {processed}%')
+                matches.append(filename)
+            else:
+                LOGGER.debug(f'No match found, searched {processed}%')
+
+    LOGGER.debug(f'Found {len(matches)} matching metadata records.')
+    return matches
+
+
+def identifier(exml):
```

Comment:
Can we add annotations here (arg types, return type, docstring description of `:param exml:`